format.go

```diff
@@ -249,21 +248,45 @@ func ReadUintFromValueReader(r ValueReader) (int, error) {
 		return 0, err
 	}
 
-	res, ok := val.(float64)
-	if !ok {
-		return 0, fmt.Errorf("expected number")
+	parseAsFloat64 := func(val interface{}) (int, error) {
+		res, ok := val.(float64)
+		if !ok {
+			return 0, fmt.Errorf("expected float64")
+		}
+
+		intVal, fracVal := math.Modf(res)
+		if fracVal != 0 {
+			return 0, fmt.Errorf("expected float64 as integer")
+		}
+
+		if intVal < 0 {
+			return 0, fmt.Errorf("expected float64 as positive integer")
+		}
+
+		return int(intVal), nil
+	}
+
+	parseAsInt := func(val interface{}) (int, error) {
+		res, ok := val.(int)
+		if !ok {
+			return 0, fmt.Errorf("expected int")
+		}
+		return res, nil
 	}
 
-	intVal, fracVal := math.Modf(res)
-	if fracVal != 0 {
-		return 0, fmt.Errorf("expected integer")
+	// Try parsing as float64 first (default number type on JSON marshalling). If it fails, try
+	// parsing as int (default type for integer numbers in Go).
+	v, err1 := parseAsFloat64(val)
+	if err1 == nil {
+		return v, nil
 	}
 
-	if intVal < 0 {
-		return 0, fmt.Errorf("expected positive integer")
+	v, err2 := parseAsInt(val)
+	if err2 == nil {
+		return v, nil
 	}
 
-	return int(intVal), nil
+	return 0, fmt.Errorf("%s or %s", err1, err2)
```

Comment:
This way of doing it will lead to some verbose errors: If you give it the number `5.4` you'll get the error "expected float64 as integer or expected integer".

I think this would read better if we do the type checking first:

```go
switch val := val.(type) {
case float64:
  intVal, fracVal := math.Modf(val)
  if fracVal != 0 {
    return 0, fmt.Errorf("expected float64 to be integer")
  }
  …
case int:
  …
default:
  return 0, fmt.Errorf("expected float64 or int")
}
``` 

Then you'll end up with only "expected float64 to be integer" and not the "or"-part.

Reply:
Looks cleaner yes, pushed.